File: main/linked-list-in-binary-tree/linked-list-in-binary-tree.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

function isSubPathAt(head, root) {
    if (!head) return true;
    
    if (!root || head.val !== root.val) return false;
    
    return isSubPathAt(head.next, root.left) ||
            isSubPathAt(head.next, root.right);
}

function hasNonemptySubPath(head, root) {
    if (!root) return false;
    
    return isSubPathAt(head, root) ||
            hasNonemptySubPath(head, root.left) ||
            hasNonemptySubPath(head, root.right);
}

/**
 * @param {ListNode} head
 * @param {TreeNode} root
 * @return {boolean}
 */
function isSubPath(head, root) {
    return !head || hasNonemptySubPath(head, root);
}

File: main/linked-list-in-binary-tree/linked-list-in-binary-tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    static bool isSubPath(const ListNode* head, const TreeNode* root) noexcept;
};

namespace {
    bool isSubPathAt(const ListNode* const head,
                     const TreeNode* const root) noexcept
    {
        if (!head) return true;
        
        if (!root || head->val != root->val) return false;
        
        return isSubPathAt(head->next, root->left)
                || isSubPathAt(head->next, root->right);
    }
    
    bool hasNonemptySubPath(const ListNode* const head,
                            const TreeNode* const root) noexcept
    {
        if (!root) return false;
        
        return isSubPathAt(head, root)
                || hasNonemptySubPath(head, root->left)
                || hasNonemptySubPath(head, root->right);
    }
}

bool Solution::isSubPath(const ListNode* const head,
                         const TreeNode* const root) noexcept
{
    return !head || hasNonemptySubPath(head, root);
}
